### PR TITLE
Linear algebra/power iteration

### PR DESCRIPTION
**Pull request by zakademic**
### **Describe your change:**
Added power iteration algorithm. Given a symmetric matrix and a random vector of same space, will find largest eigenvalue and corresponding eigenvector. Rayleigh quotient needed to compute eigenvalue given an eigenvector. 

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [ ] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [ ] This pull request is all my own work -- I have not plagiarized.
* [ ] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [ ] All filenames are in all lowercase characters with no spaces or dashes.
* [ ] All functions and variable names follow Python naming conventions.
* [ ] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by cclauss** [comment]
Is this a duplicate of
https://github.com/TheAlgorithms/Python/blob/master/linear_algebra/src/rayleigh_quotient.py

**Comment by cclauss** [comment]
@QuantumNovice Your review please.

**Comment by zakademic** [comment]
> Is this a duplicate of
> https://github.com/TheAlgorithms/Python/blob/master/linear_algebra/src/rayleigh_quotient.py

Oops I missed that when writing power iteration. The rayleigh quotient I wrote within it is a duplicate; I will remove. But power iteration is not a duplicate. Thanks!

**Comment by zakademic** [comment]
> Go through the [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md) and make sure it passes the flake8 style standard

OK will do.

File: linear_algebra/src/power_iteration.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    input_matrix: input matrix.
+    Numpy array. np.shape(input_matrix) == (N,N).
+    vector: input vector.
+    Numpy array. np.shape(vector) == (N,) or (N,1)
+    """
+
+    num = np.dot(vector.T, np.dot(input_matrix, vector))
+    den = np.dot(vector.T,vector)
+
+    return num/den
```
Comment (QuantumNovice) [inline linear_algebra/src/power_iteration.py:20]: Can you remove these lines and use this instead?
```python
from rayleigh_quotient import rayleigh_quotient
```

Comment (zakademic) [reply linear_algebra/src/power_iteration.py:20]: Working on this now, thank you!

**Comment by zakademic** [comment]
@QuantumNovice I actually removed Rayleigh quotient since in the power iteration algorithm the vectors are already normalized. This removes a dot product and division operation. 

I had not written doctests before so I referenced what was happening in Rayleigh Quotient and tried to emulate that. 

I am using VS Code now and set `black` as my auto formatter. I ran `flake8` as well. 

Thanks all for your help!

File: linear_algebra/src/power_iteration.py [hunk 2]
```diff
@@ -40,83 +23,74 @@ def power_iteration(input_matrix: np.array,vector: np.array,
...
     # or when we have small changes from one iteration to next.
 
     convergence = False
     lamda_previous = 0
     iterations = 0
     error = 1e12
 
-    while (convergence == False):
-        w = np.dot(input_matrix,vector)
+    while convergence is False:
```
Comment (cclauss) [inline linear_algebra/src/power_iteration.py:51]:
```suggestion
    while not convergence:
```

File: linear_algebra/src/power_iteration.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    Numpy array. np.shape(vector) == (N,) or (N,1)
+
+    Output
+    largest_eigenvalue: largest eigenvalue of the matrix input_matrix.
+    Float. Scalar.
+    largest_eigenvector: eigenvector corresponding to largest_eigenvalue.
+    Numpy array. np.shape(largest_eigenvector) == (N,) or (N,1).
+
+    >>> import numpy as np
+    >>> A = np.array([
```
Comment (cclauss) [inline linear_algebra/src/power_iteration.py:28]: Please use the same variable names as the function parameters so the reader does not get lost.

File: linear_algebra/src/power_iteration.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+        if error <= error_tol or iterations >= max_iterations:
+            convergence = True
+
+        lamda_previous = lamda
+
+    return lamda, vector
+
+
+def tests() -> None:
```
Comment (cclauss) [inline linear_algebra/src/power_iteration.py:73]:
```suggestion
def test_power_iteration() -> None:
```
This function name will cause pytest to find and run these tests.

File: linear_algebra/src/power_iteration.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    # Our implementation.
+    A = np.array([[41, 4, 20], [4, 26, 30], [20, 30, 50]])
+    v = np.array([41, 4, 20])
+    eigen_value, eigen_vector = power_iteration(A, v)
+
+    # Numpy implementation.
+    eigs, eigvs = np.linalg.eigh(A)
+    eig_max = eigs[-1]
+    eigv_max = eigvs[:, -1]
```
Comment (cclauss) [inline linear_algebra/src/power_iteration.py:83]: Longer variable names would improve readability for learners.

File: linear_algebra/src/power_iteration.py (hunk 3 above)
Comment (zakademic) [reply linear_algebra/src/power_iteration.py]: Thanks @cclauss this is a great point; I made the change :)

File: linear_algebra/src/power_iteration.py (hunk 5 above)
Comment (zakademic) [reply linear_algebra/src/power_iteration.py]: I made changes to be more clear and added some comments as well. Appreciate it!

**Comment by cclauss** [comment]
Nice!!

**Pull request merged**